bson/src/main/org/bson/codecs/pojo/TypeParameterMap.java

```diff
@@ -26,7 +26,7 @@
  * Maps the index of a class's generic parameter type index to a property's.
  */
 final class TypeParameterMap {
-    private final Map<Integer, Integer> propertyToClassParamIndexMap;
+    private final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap;
```

Comment:
The `TypeParameterMap` has changed from a flat level map to a nested level map.  This allows the nested generic 
types to be handled (eg: `List<List<T>>`) as we can map the levels effectively.

Essentially it allows  the mapping of:
```
List<T> => {0, 0}
List<List<T>> => {0, {0, 0}}
```